servers/rendering/renderer_rd/shaders/volumetric_fog.glsl

```diff
@@ -239,465 +108,153 @@ const vec3 halton_map[TEMPORAL_FRAMES] = vec3[](
-
-		//this causes corrupted pixels, i have no idea why..
-		if (all(bvec2(all(greaterThanEqual(position, vec3(0.0))), all(lessThan(position, voxel_gi_instances.data[i].bounds))))) {
-			position /= voxel_gi_instances.data[i].bounds;
-
-			vec4 light = vec4(0.0);
-			for (uint j = 0; j < voxel_gi_instances.data[i].mipmaps; j++) {
-				vec4 slight = textureLod(sampler3D(voxel_gi_textures[i], linear_sampler_with_mipmaps), position, float(j));
-				float a = (1.0 - light.a);
-				light += a * slight;
-			}
-
-			light.rgb *= voxel_gi_instances.data[i].dynamic_range * params.gi_inject;
-
-			total_light += light.rgb;
+	if (cull_mask > 0.0) {
+		{
+#CODE : FOG
 		}
-	}
-
-	//sdfgi
-#ifdef ENABLE_SDFGI
-
-	{
-		float blend = -1.0;
-		vec3 ambient_total = vec3(0.0);
-
-		for (uint i = 0; i < sdfgi.max_cascades; i++) {
-			vec3 cascade_pos = (world_pos - sdfgi.cascades[i].position) * sdfgi.cascades[i].to_probe;
 
-			if (any(lessThan(cascade_pos, vec3(0.0))) || any(greaterThanEqual(cascade_pos, sdfgi.cascade_probe_size))) {
-				continue; //skip cascade
-			}
-
-			vec3 base_pos = floor(cascade_pos);
-			ivec3 probe_base_pos = ivec3(base_pos);
-
-			vec4 ambient_accum = vec4(0.0);
-
-			ivec3 tex_pos = ivec3(probe_base_pos.xy, int(i));
-			tex_pos.x += probe_base_pos.z * sdfgi.probe_axis_size;
-
-			for (uint j = 0; j < 8; j++) {
-				ivec3 offset = (ivec3(j) >> ivec3(0, 1, 2)) & ivec3(1, 1, 1);
-				ivec3 probe_posi = probe_base_pos;
-				probe_posi += offset;
-
-				// Compute weight
-
-				vec3 probe_pos = vec3(probe_posi);
-				vec3 probe_to_pos = cascade_pos - probe_pos;
-
-				vec3 trilinear = vec3(1.0) - abs(probe_to_pos);
-				float weight = trilinear.x * trilinear.y * trilinear.z;
-
-				// Compute lightprobe occlusion
-
-				if (sdfgi.use_occlusion) {
-					ivec3 occ_indexv = abs((sdfgi.cascades[i].probe_world_offset + probe_posi) & ivec3(1, 1, 1)) * ivec3(1, 2, 4);
-					vec4 occ_mask = mix(vec4(0.0), vec4(1.0), equal(ivec4(occ_indexv.x | occ_indexv.y), ivec4(0, 1, 2, 3)));
-
-					vec3 occ_pos = clamp(cascade_pos, probe_pos - sdfgi.occlusion_clamp, probe_pos + sdfgi.occlusion_clamp) * sdfgi.probe_to_uvw;
-					occ_pos.z += float(i);
-					if (occ_indexv.z != 0) { //z bit is on, means index is >=4, so make it switch to the other half of textures
-						occ_pos.x += 1.0;
-					}
-
-					occ_pos *= sdfgi.occlusion_renormalize;
-					float occlusion = dot(textureLod(sampler3D(sdfgi_occlusion_texture, linear_sampler), occ_pos, 0.0), occ_mask);
-
-					weight *= max(occlusion, 0.01);
+#ifdef DENSITY_USED
+		density *= cull_mask;
+		if (abs(density) > 0.001) {
+			int final_density = int(density * DENSITY_SCALE);
+			imageAtomicAdd(density_only_map, pos, uint(final_density));
+
+#ifdef EMISSION_USED
+			{
+				emission *= clamp(density, 0.0, 1.0);
+				emission = max(vec3(0.0), emission);
+				// Scale them to 8 and 7 bits so we have an effective range of 0-8
+				uvec3 emission_u = uvec3(emission.r * 255.0, emission.g * 255.0, emission.b * 127.0);
```

Comment:
I suspect this is causing the yellowing effect with low density values. (I got it to show up again.)
Setting `DENSITY` to 0.00787 (just below 1/127) truncates the blue component to zero, but leaves R and G.

Perhaps non-linearly scaling emission would make this effect no longer noticeable.

Reply:
Interesting, I'm surprised the yellowing is so apparent. They use varying resolution because the underlying data is R11_G11_B10. So to get them all to have an effective range of 0-8, we need to to use 8/8/7 bits respectively. 

I'll try scaling emission. The other alternative is to ignore 2 bits and use 10 bits for each channel. This reduces precision in red and green, but will keep everything uniform

Reply:
Okay, scaling did the trick! Should be good now. 

I kept Emission in the 0-8 range. But I changed scattering to 0-1 as scattering should not add light to a scene.

Reply:
@clayjohn Something I just realized is that this nonlinear scaled emission is no longer mathematically correct in addition. (e.g. 0.5+0.9=1.4 becomes 0.5+0.9 ~= 3.1.) As a result, combined emissive volumes look far too bright. For accurate addition, you'd have to load, unscale, add, rescale, then store. There's no reasonable way this is possible with atomic image operations.

The alternative _reduced-bits_ option is unfortunate because 7 bits per color yields significant color inaccuracy at low fog densities. Setting fog density below 1/127, instead of yellowing, would truncate all emission. This makes light clouds look, honestly, terrifying.

The only ways around the imprecision problem that I can think of are:
A. Reduce the emission range
B. Add _another_ image3D to provide more bits
C. Add a fog volume density linear scale property to the environment, applied in-place of the current nonlinear scaling. This would effectively give the user control over the range of precision.

I think option C is the best one, but that could be done in another PR.

Reply:
@briansemrau I have another option for you to consider.

D. Use emission_energy in the environment to  scale all emissive colors and force emission in the shaders to cap at 1 (e.g. we would be defining emission color in the volume and emission energy globally)

The downside is that emission values use a fixed scale. But that isn't a big deal, you can just divide by the scale in your shader.